helix-view/src/editor.rs

```diff
@@ -398,18 +409,64 @@ impl Editor {
             .selections
             .remove(&id);
 
-        if close_buffer {
-            // get around borrowck issues
-            let doc = &self.documents[&view.doc];
+        self.tree.remove(id);
+        self._refresh();
+    }
 
-            if let Some(language_server) = doc.language_server() {
-                tokio::spawn(language_server.text_document_did_close(doc.identifier()));
-            }
-            self.documents.remove(&view.doc);
+    pub fn close_document(&mut self, doc_id: DocumentId, force: bool) -> anyhow::Result<()> {
+        let doc = match self.documents.get(&doc_id) {
+            Some(doc) => doc,
+            None => anyhow::bail!("document does not exist"),
+        };
+
+        if !force && doc.is_modified() {
+            anyhow::bail!(
+                "buffer {:?} is modified",
+                doc.relative_path()
+                    .map(|path| path.to_string_lossy().to_string())
+                    .unwrap_or_else(|| "[scratch]".into())
+            );
+        }
+
+        if let Some(language_server) = doc.language_server() {
+            tokio::spawn(language_server.text_document_did_close(doc.identifier()));
+        }
+
+        let views_to_close = self
+            .tree
+            .views()
+            .filter_map(|(view, focus)| {
+                if view.doc == doc_id {
+                    Some((view.id, focus))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        for (view_id, _focus) in views_to_close {
+            self.close(view_id);
+        }
```

Comment:
```suggestion
        let views_to_close = self
            .tree
            .views()
            .filter(|(view, focus)| view.doc == doc_id);

        for (view, _) in views_to_close {
            self.close(view.id);
        }
```
This should make it simpler. Also, I think we should have a helper to find view by document id.

Reply:
Unfortunately, that trips up the borrow checker:

```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
   --> helix-view/src/editor.rs:441:13
    |
435 |           let views_to_close = self
    |  ______________________________-
436 | |             .tree
    | |_________________- immutable borrow occurs here
...
440 |           for (view, _focus) in views_to_close {
    |                                 -------------- immutable borrow later used here
441 |               self.close(view.id);
    |               ^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here

For more information about this error, try `rustc --explain E0502`.
error: could not compile `helix-view` due to previous error
```